### PR TITLE
Hcal simple reconstructor with method3

### PR DESCRIPTION
**Pull request by igv4321**
Including "method 3" initialization in the HcalSimpleReconstructor module.

Making a separate python cff for method 3 configuration and importing this configuration into HcalHitReconstructor and HcalSimpleReconstructor configs.

### CODE REVIEW

File: RecoLocalCalo/HcalRecProducers/python/HcalHitReconstructor_hbhe_cfi.py [hunk 1]
```diff
@@ -131,11 +132,9 @@
...
     # Configuration parameters for Method 3
-    pedestalSubtractionType = cms.int32(1),
-    pedestalUpperLimit      = cms.double(2.7),
-    timeSlewParsType        = cms.int32(1), # 0: TestStand, 1:Data, 2:MC, 3:InputPars. Parametrization function is par0 + par1*log(fC+par2).
-    timeSlewPars            = cms.vdouble( 12.2999, -2.19142, 0, 12.2999, -2.19142, 0, 12.2999, -2.19142, 0 ),# HB par0, HB par1, HB par2, BE par0, BE par1, BE par2, HE par0, HE par1, HE par2
-    respCorrM3              = cms.double(0.95)# This factor is used to align the the Method3 with the Method2 response
-    )
-
-
+    pedestalSubtractionType = cms.int32(method3.pedestalSubtractionType),
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/python/HcalHitReconstructor_hbhe_cfi.py]: I think you can do plain assignment if you define the types in HBHEMethod3Parameters, without this "cast".

**Comment by igv4321** [comment]
Yes, indeed. This is also what I tried earlier, but it did not work for me initially. Probably had an extra comma somewhere of something like that.

File: RecoLocalCalo/HcalRecProducers/python/HBHEMethod3Parameters_cfi.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+
+# Configuration parameters for Method 3
+pedestalSubtractionType = cms.int32(1)
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/python/HBHEMethod3Parameters_cfi.py]: you can wrap this in a PSet and then include it in the module (has to be included before any named parameters of the module
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideAboutPythonConfigFile#Parameter_Set_PSet_Objects
)
https://github.com/cms-sw/cmssw/compare/cms-sw:CMSSW_7_6_X...slava77:CMSSW_7_6_0_pre7/sign613/alt-11769

File: RecoLocalCalo/HcalRecProducers/python/HcalHitReconstructor_ho_cfi.py [hunk 3]
```diff
@@ -31,7 +32,11 @@
     setTimingTrustFlags = cms.bool(False), # timing flags currently only implemented for HF
     setPulseShapeFlags = cms.bool(False),  # not yet defined for HO
     setNegativeFlags          = cms.bool(False),  # only in HBHE
-    saturationParameters=  cms.PSet(maxADCvalue=cms.int32(127))
+    saturationParameters=  cms.PSet(maxADCvalue=cms.int32(127)),
+    # Configuration parameters for Method 3
+    pedestalSubtractionType = method3.pedestalSubtractionType,
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/python/HcalHitReconstructor_ho_cfi.py]: the framework will put the default values defined with fillDescription from the HcalHitReconstructor .
The parameters are bogus anyways (nothing to be set for HF or HO hits) as an aux energy setting parameters.

Even if we will have M3 for HO or HF, I doubt very much that the HBHE-derived values will apply.
I suggest to remove to reduce confusion.

File: RecoLocalCalo/HcalRecProducers/src/HcalSimpleReconstructor.cc [hunk 4]
```diff
@@ -25,6 +25,14 @@ HcalSimpleReconstructor::HcalSimpleReconstructor(edm::ParameterSet const& conf):
...
   theTopology(0)
 {
+  // Intitialize "method 3"
+  reco_.setMeth3Params(
+            conf.getParameter<int>     ("pedestalSubtractionType"),
+            conf.getParameter<double>  ("pedestalUpperLimit"),
+            conf.getParameter<int>     ("timeSlewParsType"),
+            conf.getParameter<std::vector<double> >("timeSlewPars"),
+            conf.getParameter<double>  ("respCorrM3")
+      );
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/HcalSimpleReconstructor.cc]: similar to HcalHitReconstructor, please add fillDescriptions method

**Comment by igv4321** [comment]
OK, fillDescriptions added

**Comment by slava77** [comment]
@cmsbuild please test
[unsure yet if the IB is alive]

**Comment by slava77** [comment]
As discussed in https://github.com/cms-sw/cmssw/issues/11735
more strict M3 initialization strategy is desired (currently, construction and initialization are done in separate methods)
One compile time choice is proposed here
https://github.com/slava77/cmssw/commit/62449ed

Another alternative, less desired IMO, is to do a run-time check

**Comment by smuzaffar** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11769	37620ae
- resolves uninitialized conditions/undefined behavior reported in https://github.com/cms-sw/cmssw/issues/11735 . A more robust solution forcing initialization at construction (without delays to special init function) was discussed as something to be done reasonably soon.
- jenkins tests pass and comparisons with the baseline show no difference

**Comment by davidlange6** [comment]
+1

**Pull request merged**